docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+- continuing to develop in the kubernetes/kubernetes repo results in more *bad* dependencies being added
+  even as we try to remove old ones
+- many commands depend on test infrastructure bespoke to kubernetes/kubernetes, which would need to be moved as well
+
+Additionally, many kubectl commands are laden with technical debt, using anti-patterns for working with APIs
+that do not work with version skew or support extensibility.  We have since grown out of using these patterns,
+but they are still pervasive.  Frequently, is it much faster and effective to rewrite large pieces
+instead of trying to refactor them into different designs.
+
+Some examples of anti-patterns:
+
+- Relying on specific compiled in API types for generic operations (scale, rollback)
+- Using internal type definitions (printers, various other commands)
+- Round tripping objects read from the server through go structures (some exceptions where this maybe ok)
+
+## Goals for ktl
+
+Thing we need.
+
+Keep the advantages of:
+
+- easy distribution
+- reasonably sized (~100MB)
+- easy installation
+- use common cli / client infrastructure
+- discoverable commands
+
+And also:
+
+- allow the cli to be released independently from the cluster
+- allow sub command groups within the cli to be released independently from one another
+- allow a decentralized ecosystem of tools to leverage centralized maintained cli / client infrastructure
+- facilitate end-to-end ownership of the cli, and in some cases sub command groups with the cli
+- facilitate decentralized development of extensions for the cli
+
+## Anti-goals
+
+Things we want to avoid.
+
+- block on moving existing kubectl commands out of the kubernetes/kubernetes repo
+- rewrite kubectl from the ground up in a new repo
+
+## Non-goals
+
+Even if these are good ideas, don't let them distract us from meeting our goals will simpler solutions.
+
+- build solution for discovering installable plugins and installing them
+  - rely on existing package management solutions for this until we need something more
+- invent new build and distribution infrastructure
+- fix issues with the existing kubectl commands
+- dog fooding the plugin mechanism for core commands
+
+## Reference-level explanation
+
+- build a new cli binary `ktl` (kubectl2) under the `kubernetes/ktl` repo that dispatches to commands developed in other repos
+- keep old cli commands in `kubernetes/kubernetes/cmd/kubectl` and vendor them into `kubectl/ktl`
+- build new cli commands in `kubernetes/kubectl` and vendor them into `kurnetes/ktl`
+- build common cli Kubernetes client infrastructure and libraries that can be used to develop a decentralized cli ecosystem
+
+### ktl
+
+In a new repo (`kubernetes/ktl`), create the `ktl` binary that dispatches (both statically and dynamically)
+to commands developed in other repos.
+
+#### Dispatch
+
+Static dispatch:
+
+- vendor in kubernetes/kubernetes/cmd/kubectl cobra commands, and register them under `ktl kube`
```

Comment:
If you were going to do this (I'm not weighing in on that one way or the other yet), why not be more judicious about those things included.  If you made `kubectl` match the dynamic plugin mechanism, you could avoid ever taking a dependency on `k8s.io/kubernetes`.  Doing this appears to be in opposition to your previous statement about "many kubectl commands depend on internal kubernetes/kubernetes libraries".   You'd still have that dependency, it would just be harder to see.

Reply:
Using the dynamic plugin mechanism complicates things like distribution and auto-update.  See goals for more details.

This would quarantine the dependency without requiring a ground up rewrite.  That is sufficient for our immediate goals.

Reply:
> This would quarantine the dependency without requiring a ground up rewrite. That is sufficient for our immediate goals.

It usually doesn't result in that quarantine.  See https://github.com/kubernetes/kubectl/pull/179#discussion_r158068798 .  We've tried on very simple vendorings of client-go and ended up with intractable failures.

Reply:
Throwing it behind an *internal* package should be sufficient.

Reply:
The "glog does bad things to global flags package" is I think a stronger argument against vendoring.

Reply:
I can't see the benefit of this - *every* system that has ktl is going to have kubectl for various reasons.  Distribute them together, have ktl use dynamic dispatch until such a time as you can break it up.  I can't imagine someone who is going to have just ktl installed.